File: services/config/providers/local.ts
/**
 * Mossen 本地 provider 实现 (G1-2):
 *  - LocalDefaultProvider: 内置默认值, 不可写, priority 最低
 *  - UserSettingsProvider:    ~/.mossen/settings.json (MOSSEN_CONFIG_DIR 可 override)
 *  - ProjectSettingsProvider: <cwd>/.mossen/settings.json
 *
 * 设计要点:
 *  - 不缓存; 每次 get 直接读盘 (G1-2 简单实现, G2 阶段可以加 mtime 缓存)
 *  - settings.json 写入失败抛错 (set 是 mutation, 不能静默吞)
 *  - 文件不存在 / 解析失败 → get 返回 undefined (provider miss)
 *  - clear 不存在的文件 = no-op
 */

import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'

import {
  PROVIDER_PRIORITY,
  type ConfigValueSource,
  type MossenConfigProvider,
  type ProviderResult,
} from '../types.js'
import { MOSSEN_BUILTIN_DEFAULTS } from '../defaults.js'

/** 内置默认值, 不可写 */
export class LocalDefaultProvider implements MossenConfigProvider {
  readonly name: ConfigValueSource = 'default'

  readonly priority = PROVIDER_PRIORITY.default

  readonly enabled = true

  get<T>(key: string): ProviderResult<T> {
    if (Object.prototype.hasOwnProperty.call(MOSSEN_BUILTIN_DEFAULTS, key)) {
      return {
        value: MOSSEN_BUILTIN_DEFAULTS[key] as T,
        source: 'default',
      }
    }
    return undefined
  }
}

/** SettingsProvider 共享实现 (User / Project 都基于此) */
abstract class SettingsProviderBase implements MossenConfigProvider {
  abstract readonly name: ConfigValueSource

  abstract readonly priority: number

  readonly enabled = true

  /** 子类返回 settings.json 完整路径 */
  protected abstract resolveSettingsPath(): string

  /**
   * 子类可 override; 返回 mode (如 0o600) 表示写入后强制 chmod 到该权限.
   * 返回 null 表示沿用 OS umask 默认.
   * User scope 覆盖为 0o600 因 settings.json 内嵌 apiKey (D-S09-1=A).
   */
  protected getSecurePermissionMode(): number | null {
    return null
  }

  get<T>(key: string): ProviderResult<T> {
    const data = this.readSettings()
    if (data && Object.prototype.hasOwnProperty.call(data, key)) {
      return { value: data[key] as T, source: this.name }
    }
    return undefined
  }

  set<T>(key: string, value: T): void {
    const settingsPath = this.resolveSettingsPath()
    const current = this.readSettings() ?? {}
    current[key] = value as unknown
    fs.mkdirSync(path.dirname(settingsPath), { recursive: true })
    fs.writeFileSync(
      settingsPath,
      `${JSON.stringify(current, null, 2)}\n`,
      'utf-8',
    )
    this.enforceSecurePermission(settingsPath)
  }

  clear(key?: string): void {
    const settingsPath = this.resolveSettingsPath()
    if (!fs.existsSync(settingsPath)) return
    if (key === undefined) {
      try {
        fs.unlinkSync(settingsPath)
      } catch {
        // ignore — 文件可能已被并发删除
      }
      return
    }
    const current = this.readSettings()
    if (current && Object.prototype.hasOwnProperty.call(current, key)) {
      delete current[key]
      fs.writeFileSync(
        settingsPath,
        `${JSON.stringify(current, null, 2)}\n`,
        'utf-8',
      )
      this.enforceSecurePermission(settingsPath)
    }
  }

  private enforceSecurePermission(settingsPath: string): void {
    const securePerm = this.getSecurePermissionMode()
    if (securePerm === null) return
    try {
      const currentMode = fs.statSync(settingsPath).mode & 0o777
      if (currentMode !== securePerm) {
        fs.chmodSync(settingsPath, securePerm)
      }
    } catch {
      // chmod 失败不阻塞写入 (如 Windows / 不支持 mode 的文件系统).
    }
  }

  private readSettings(): Record<string, unknown> | undefined {
    const settingsPath = this.resolveSettingsPath()
    if (!fs.existsSync(settingsPath)) return undefined
    try {
      const raw = fs.readFileSync(settingsPath, 'utf-8')
      const parsed = JSON.parse(raw) as unknown
      if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
        return parsed as Record<string, unknown>
      }
      return undefined
    } catch {
      return undefined
    }
  }
}

/** ~/.mossen/settings.json (MOSSEN_CONFIG_DIR env 可 override 测试) */
export class UserSettingsProvider extends SettingsProviderBase {
  readonly name: ConfigValueSource = 'user'

  readonly priority = PROVIDER_PRIORITY.user

  protected resolveSettingsPath(): string {
    const configDir =
      process.env.MOSSEN_CONFIG_DIR ?? path.join(os.homedir(), '.mossen')
    return path.join(configDir, 'settings.json')
  }

  protected getSecurePermissionMode(): number | null {
    return 0o600
  }
}

/** <cwd>/.mossen/settings.json */
export class ProjectSettingsProvider extends SettingsProviderBase {
  readonly name: ConfigValueSource = 'project'

  readonly priority = PROVIDER_PRIORITY.project

  protected resolveSettingsPath(): string {
    return path.join(process.cwd(), '.mossen', 'settings.json')
  }
}

File: crates/mossen-agent/src/services/config/providers/local.rs
//! Local file-based configuration providers.
//!
//! - LocalDefaultProvider: built-in defaults, read-only, lowest priority
//! - UserSettingsProvider: ~/.mossen/settings.json
//! - ProjectSettingsProvider: <cwd>/.mossen/settings.json

use parking_lot::Mutex;
use serde_json::Value;
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};

use crate::services::config::defaults::MOSSEN_BUILTIN_DEFAULTS;
use crate::services::config::types::{ConfigValueSource, MossenConfigProvider, ProviderResult};

/// Built-in default provider (read-only, lowest priority).
pub struct LocalDefaultProvider;

impl MossenConfigProvider for LocalDefaultProvider {
    fn name(&self) -> ConfigValueSource {
        ConfigValueSource::Default
    }

    fn priority(&self) -> u8 {
        ConfigValueSource::Default.priority()
    }

    fn enabled(&self) -> bool {
        true
    }

    fn get(&self, key: &str) -> Option<ProviderResult> {
        MOSSEN_BUILTIN_DEFAULTS.get(key).map(|value| ProviderResult {
            value: value.clone(),
            source: ConfigValueSource::Default,
            resolved_key: None,
        })
    }
}

/// Shared base implementation for file-based settings providers.
struct SettingsProviderInner {
    settings_path: PathBuf,
    secure_permission_mode: Option<u32>,
}

impl SettingsProviderInner {
    fn read_settings(&self) -> Option<HashMap<String, Value>> {
        if !self.settings_path.exists() {
            return None;
        }
        let raw = fs::read_to_string(&self.settings_path).ok()?;
        let parsed: Value = serde_json::from_str(&raw).ok()?;
        match parsed {
            Value::Object(map) => {
                let mut result = HashMap::new();
                for (k, v) in map {
                    result.insert(k, v);
                }
                Some(result)
            }
            _ => None,
        }
    }

    fn get_value(&self, key: &str, source: ConfigValueSource) -> Option<ProviderResult> {
        let data = self.read_settings()?;
        data.get(key).map(|value| ProviderResult {
            value: value.clone(),
            source,
            resolved_key: None,
        })
    }

    fn set_value(&self, key: &str, value: Value) {
        let mut current = self.read_settings().unwrap_or_default();
        current.insert(key.to_string(), value);
        if let Some(parent) = self.settings_path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        let map: serde_json::Map<String, Value> = current.into_iter().collect();
        let json_str = serde_json::to_string_pretty(&Value::Object(map)).unwrap_or_default();
        let content = format!("{}\n", json_str);
        let _ = fs::write(&self.settings_path, content);
        self.enforce_secure_permission();
    }

    fn clear_value(&self, key: Option<&str>) {
        if !self.settings_path.exists() {
            return;
        }
        match key {
            None => {
                let _ = fs::remove_file(&self.settings_path);
            }
            Some(k) => {
                if let Some(mut current) = self.read_settings() {
                    if current.contains_key(k) {
                        current.remove(k);
                        let map: serde_json::Map<String, Value> =
                            current.into_iter().collect();
                        let json_str =
                            serde_json::to_string_pretty(&Value::Object(map)).unwrap_or_default();
                        let content = format!("{}\n", json_str);
                        let _ = fs::write(&self.settings_path, content);
                        self.enforce_secure_permission();
                    }
                }
            }
        }
    }

    #[cfg(unix)]
    fn enforce_secure_permission(&self) {
        use std::os::unix::fs::PermissionsExt;
        if let Some(mode) = self.secure_permission_mode {
            if let Ok(metadata) = fs::metadata(&self.settings_path) {
                let current_mode = metadata.permissions().mode() & 0o777;
                if current_mode != mode {
                    let _ = fs::set_permissions(
                        &self.settings_path,
                        fs::Permissions::from_mode(mode),
                    );
                }
            }
        }
    }

    #[cfg(not(unix))]
    fn enforce_secure_permission(&self) {
        // chmod not available on non-Unix platforms
    }
}

/// User settings provider (~/.mossen/settings.json).
pub struct UserSettingsProvider {
    inner: Mutex<SettingsProviderInner>,
}

impl UserSettingsProvider {
    pub fn new() -> Self {
        let config_dir = env::var("MOSSEN_CONFIG_DIR").unwrap_or_else(|_| {
            let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
            home.join(".mossen").to_string_lossy().to_string()
        });
        let settings_path = Path::new(&config_dir).join("settings.json");
        Self {
            inner: Mutex::new(SettingsProviderInner {
                settings_path,
                secure_permission_mode: Some(0o600),
            }),
        }
    }
}

impl Default for UserSettingsProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl MossenConfigProvider for UserSettingsProvider {
    fn name(&self) -> ConfigValueSource {
        ConfigValueSource::User
    }

    fn priority(&self) -> u8 {
        ConfigValueSource::User.priority()
    }

    fn enabled(&self) -> bool {
        true
    }

    fn get(&self, key: &str) -> Option<ProviderResult> {
        let inner = self.inner.lock();
        inner.get_value(key, ConfigValueSource::User)
    }

    fn set(&self, key: &str, value: Value) {
        let inner = self.inner.lock();
        inner.set_value(key, value);
    }

    fn clear(&self, key: Option<&str>) {
        let inner = self.inner.lock();
        inner.clear_value(key);
    }
}

/// Project settings provider (<cwd>/.mossen/settings.json).
pub struct ProjectSettingsProvider {
    inner: Mutex<SettingsProviderInner>,
}

impl ProjectSettingsProvider {
    pub fn new() -> Self {
        let cwd = env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
        let settings_path = cwd.join(".mossen").join("settings.json");
        Self {
            inner: Mutex::new(SettingsProviderInner {
                settings_path,
                secure_permission_mode: None,
            }),
        }
    }
}

impl Default for ProjectSettingsProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl MossenConfigProvider for ProjectSettingsProvider {
    fn name(&self) -> ConfigValueSource {
        ConfigValueSource::Project
    }

    fn priority(&self) -> u8 {
        ConfigValueSource::Project.priority()
    }

    fn enabled(&self) -> bool {
        true
    }

    fn get(&self, key: &str) -> Option<ProviderResult> {
        let inner = self.inner.lock();
        inner.get_value(key, ConfigValueSource::Project)
    }

    fn set(&self, key: &str, value: Value) {
        let inner = self.inner.lock();
        inner.set_value(key, value);
    }

    fn clear(&self, key: Option<&str>) {
        let inner = self.inner.lock();
        inner.clear_value(key);
    }
}
